### PR TITLE
Apply policy resource restrictions for file extensions

### PR DESCRIPTION
**Pull request by prakashsvmx**
Fixes https://github.com/minio/console/issues/2838

Sample policy to test:
( Allow and deny both actions are used)

<details>
<summary> Policy 1</summary>

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:ListBucket",
                "s3:DeleteObject",
                "s3:GetObject",
                "s3:GetObjectVersion",
                "s3:GetObjectVersionTagging",
                "s3:ListAllMyBuckets"
            ],
            "Resource": [
                "arn:aws:s3:::mybucket/*"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:PutObject"
            ],
            "Resource": [
                "arn:aws:s3:::mybucket/*.jpg",
                "arn:aws:s3:::mybucket/anything/*",
                "arn:aws:s3:::mybucket/jpg/*.jpg",
                "arn:aws:s3:::mybucket/l1/l2/*.txt",
                "arn:aws:s3:::mybucket/png/*.png",
                "arn:aws:s3:::mybucket/txt/*.doc",
                "arn:aws:s3:::mybucket/txt/*.txt"
            ]
        }
    ]
}
```
</details>

<details>

<summary> Policy 2</summary>

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:DeleteObject",
                "s3:GetObject",
                "s3:GetObjectVersion",
                "s3:GetObjectVersionTagging",
                "s3:ListAllMyBuckets",
                "s3:ListBucket"
            ],
            "Resource": [
                "arn:aws:s3:::mybucket/*"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:PutObject"
            ],
            "Resource": [
                "arn:aws:s3:::mybucket/txt/*.txt",
                "arn:aws:s3:::mybucket/*.jpg",
                "arn:aws:s3:::mybucket/anything/*",
                "arn:aws:s3:::mybucket/jpg/*.jpg",
                "arn:aws:s3:::mybucket/l1/l2/*.txt",
                "arn:aws:s3:::mybucket/png/*.png",
                "arn:aws:s3:::mybucket/txt/*.doc"
            ]
        },
        {
            "Effect": "Deny",
            "Action": [
                "s3:PutObject"
            ],
            "Resource": [
                "arn:aws:s3:::mybucket/txt/*.doc",
                "arn:aws:s3:::mybucket/txt/*.txt"
            ]
        }
    ]
}
```

</details>

Testing:

`CI=true  MINIO_ROOT_USER=minio MINIO_ROOT_PASSWORD=minio123 minio server /tmp/site1{1...4}  --address ":22000" --console-address ":9025" `

`CONSOLE_DEV_MODE=on CONSOLE_MINIO_SERVER=http://localhost:22000 ./console server`

- Allow/Deny file extensions on a root level,
- Allow/Deny at prefix level
- Drag and Drop individual files
- Drag and drop folders ( only allowed extensions are uploaded )

### CODE REVIEW

File: portal-ui/src/screens/Console/Buckets/ListBuckets/UploadPermissionUtils.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { ISessionPermissions } from "../../types";
```
Comment (harshavardhana) [inline portal-ui/src/screens/Console/Buckets/ListBuckets/UploadPermissionUtils.ts:1]: license header missing

Comment (prakashsvmx) [reply portal-ui/src/screens/Console/Buckets/ListBuckets/UploadPermissionUtils.ts:17]: Thank you . I have updated.

**Review by bexsoft — CHANGES_REQUESTED** [review]
There is an issue if I try to upload a file using drag and drop:

**Comment by prakashsvmx** [comment]
Thank you @bexsoft  :+1: . I have added logs and improved the dnd handling to accept one or more files.

**Comment by dvaldivia** [comment]
there's a warning @prakashsvmx

**Comment by prakashsvmx** [comment]
@dvaldivia thank you, i have updated the type reference.

**Review by bexsoft — CHANGES_REQUESTED** [review]
If I try to upload a file with the extension `.JPG` (Uppercase), an error is thrown

Screenshot 2023-06-14 at 19 17 59

Trying to do drag and drop, I see the following issue in any case:

Screenshot 2023-06-14 at 19 18 47

If I try to create a new subpath inside anything folder, button is not enabled
Screenshot 2023-06-14 at 19 26 28

**Comment by prakashsvmx** [comment]
@bexsoft 
> 1. If I try to upload a file with the extension .JPG (Uppercase), an error is thrown

it is working as expected. test with mc

> 2. Trying to do drag and drop, I see the following issue in any case:
  
 Fixed
   
>3. If I try to create a new subpath inside anything folder, button is not enabled

it is working as per master/release.  ( in master/release as well it is disabled)


<details>

```
➜ mc cp 1.png fxt/mybucket/png                 
...-file-types/1.png: 4 B / 4 B ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 501 B/s 0s 

# Copy PNG ( uppercase extension)

 ➜ mc cp 2.PNG fxt/mybucket/png/       
mc: <ERROR> Failed to copy `/home/prakash/Downloads/test-all-file-types/2.PNG`. Insufficient permissions to access this path `http://localhost:22000/mybucket/png/2.PNG`

 ➜ mc ls fxt/mybucket/png/          
[2023-06-15 09:48:22 IST]     4B STANDARD 1.png
[2023-06-15 08:50:22 IST] 9.5KiB STANDARD image (4).png
```
</details>

**Review by bexsoft — APPROVED** [review]
LGTM

**Review by dvaldivia — APPROVED** [review]

**Review by cesnietor — CHANGES_REQUESTED** [review]
reviewing

**Comment by cesnietor** [comment]
@prakashsvmx  I was also expecing a similar thing as @bexsoft.
And I think we should solve the following:
Since JPG (uppercase) is not allowed, it shouldn't allow you to select the file in the browser, similar to how others are not allowed, otherwise it's letting me know that it's allowed but then the app fails. This however can be an enhancement.

Screenshot 2023-06-15 at 10 03 57 AM

**Review by cesnietor — APPROVED** [review]
LGTM

**Comment by prakashsvmx** [comment]
Thank you @cesnietor  :+1:

**Pull request merged**